fem/coefficient.cpp

```diff
@@ -758,4 +758,96 @@ double ComputeGlobalLpNorm(double p, VectorCoefficient &coeff, ParMesh &pmesh,
 }
 #endif
 
+VectorQuadratureFunctionCoefficient::VectorQuadratureFunctionCoefficient(
+   QuadratureFunction *qf)
+   : VectorCoefficient(qf->GetVDim())
+{
+   QuadF = qf;
+   index = 0;
+   length = qf->GetVDim();
+}
+
+void VectorQuadratureFunctionCoefficient::SetQuadratureFunction(
+   QuadratureFunction *qf)
+{
+   index = 0;
+   length = qf->GetVDim();
+   vdim = length;
+   QuadF = qf;
+}
+
+void VectorQuadratureFunctionCoefficient::SetLength(int _length)
+{
+   MFEM_VERIFY(_length > 0, "Length must be > 0");
+
+   int diff = QuadF->GetVDim() - index;
+   MFEM_VERIFY(_length <= diff,
+               "Length must be <= (QuadratureFunction length - index)");
+
+   length = _length;
+   vdim = length;
+}
+
+void VectorQuadratureFunctionCoefficient::SetIndex(int _index)
+{
+   MFEM_VERIFY(_index >= 0, "Index must be >= 0");
+   MFEM_VERIFY(_index < QuadF->GetVDim(),
+               "Index must be < the QuadratureFunction length");
+   index = _index;
+   // check to see if length needs to be modified
+   int diff = QuadF->GetVDim() - index;
+   if (length > diff)
+   {
+      length = diff;
+      vdim = length;
+   }
+}
+
+void VectorQuadratureFunctionCoefficient::Eval(Vector &V,
+                                               ElementTransformation &T,
+                                               const IntegrationPoint &ip)
+{
+   QuadF->HostRead();
```

Comment:
Looks like you are mixing this with GPU code, is that right?

Reply:
It can be mixed with GPU code. At least in ExaConstit, we have the ```QuadratureFunctions``` evaluated\evolved on the GPU, and then we use this class to project things to a GridFunction for post-processing/visualization needs which currently requires everything to be back on the host. Does this help answer your question?